### PR TITLE
RUBY-2766 Allow hint for unacknowledged writes using OP_MSG when  supported by the server

### PR DESCRIPTION
**Pull request by Neilshweky**
The PR for RUBY-2973 needs to be reviewed first: https://github.com/mongodb/mongo-ruby-driver/pull/2465

### CODE REVIEW

File: lib/mongo/collection/view/writable.rb [hunk 1]
```diff
@@ -93,9 +95,11 @@ def find_one_and_delete(opts = {})
                 txn_num: txn_num,
               ).execute_with_connection(connection, context: context)
             end
-          end.first['value']
+          end.first&.fetch('value', nil)
```
Comment (p-mongo) [inline lib/mongo/collection/view/writable.rb:98]: An amusing/cute way of spelling this that retains the `[]` operator is `first&.[]('value')`.

File: spec/runners/unified/test.rb [hunk 2]
```diff
@@ -333,8 +333,10 @@ def execute_operation(op)
         else
           result = send(method_name, op)
           if expected_result = op.use('expectResult')
-            if result.nil? && !expected_result.empty?
-              raise Error::ResultMismatch, "Actual result nil but expected result #{expected_result}"
+            if result.nil? && expected_result.key?("$$unsetOrMatches")
```
Comment (p-mongo) [inline spec/runners/unified/test.rb:336]: This needs to assert that unsetormatches value is truthy.

File: spec/mongo/bulk_write_spec.rb [hunk 3]
```diff
@@ -257,10 +257,24 @@
...
-                expect do
-                  bulk_write.execute
-                end.to raise_error(Mongo::Error::UnsupportedOption, /The hint option cannot be specified on an unacknowledged write operation/)
+              context "on 4.4+ servers" do
+                min_server_version '4.4'
+
+                it 'raises a client-side error' do
+                  expect do
+                    bulk_write.execute
+                  end.to_not raise_error(Mongo::Error::UnsupportedOption)
```
[the remark this answers is not in the document]
Comment (Neilshweky) [reply spec/mongo/bulk_write_spec.rb:266]: yup my bad

File: spec/runners/unified/assertions.rb [hunk 4]
```diff
@@ -163,8 +163,12 @@ def assert_eq(actual, expected, msg)
     end
 
     def assert_matches(actual, expected, msg)
-      if actual.nil? && !expected.nil?
-        raise Error::ResultMismatch, "#{msg}: expected #{expected} but got nil"
+      if actual.nil?
+        if expected_result.key?("$$unsetOrMatches")
```
[the remark this answers is not in the document]
Comment (Neilshweky) [reply spec/runners/unified/assertions.rb:167]: I don't think so... This portion of the code is just dealing with what happens when the actual is `nil`. If the expected is not `nil` then we automatically want to fail, however if the expected is an `$$unsetOrMatches` then we're okay with actual being `nil` and we just return as a match.

File: spec/runners/unified/assertions.rb [hunk 5]
```diff
@@ -216,23 +225,30 @@ def assert_matches(actual, expected, msg)
...
       when 'double'
         Float === object
+      when 'string'
+        String === object
       else
         raise NotImplementedError, "Unhandled type #{type}"
       end
-      unless ok
-        raise Error::ResultMismatch, "Object #{object} is not of type #{type}"
-      end
```
[the remark this answers is not in the document]
Comment (Neilshweky) [reply spec/runners/unified/assertions.rb:235]: done

File: spec/runners/unified/test.rb (hunk 2 above)
Comment (Neilshweky) [reply spec/runners/unified/test.rb:336]: Again I'm not sure that's correct as per my comments above. I think we might have different understandings about how unsetOrMatches works.

File: lib/mongo/bulk_write.rb [hunk 6]
```diff
@@ -276,12 +276,42 @@ def validate_array_filters!(connection)
...
+    # and for the rest, the client can only send it for 4.4+:
+    #   - updateOne
+    #   - updateMany
+    #   - replaceOne
+    #
+    # @param [ Connection ] connection The connection object.
+    #
+    # @return [ true | false ] Whether the request is able to send hints for
+    #   the current server version.
+    def can_hint(connection)
```
[the remark this answers is not in the document]
Comment (Neilshweky) [reply lib/mongo/bulk_write.rb:300]: done

**Review by p-mongo — APPROVED** [review]

**Pull request merged**